File: core/key_manager.py
"""PrxyClaude · Key Manager (round-robin with ban-on-429)"""

from __future__ import annotations

import time
from dataclasses import dataclass

from loguru import logger

BAN_DURATION_429 = 300.0
MAX_ERROR_RATE = 0.5


@dataclass
class KeySlot:
    key: str
    usage_count: int = 0
    error_count: int = 0
    last_used_at: float | None = None
    banned_until: float | None = None
    rate_limit_hits: int = 0
    consecutive_failures: int = 0


_pools: dict[str, list[KeySlot]] = {}
_round_robin_index: dict[str, int] = {}


def add_key(provider_id: str, key: str) -> None:
    if provider_id not in _pools:
        _pools[provider_id] = []
        _round_robin_index[provider_id] = 0
    if not any(s.key == key for s in _pools[provider_id]):
        _pools[provider_id].append(KeySlot(key=key))
        logger.info(f"[keys] Added key for {provider_id}")


def get_next_key(provider_id: str) -> str | None:
    slots = _pools.get(provider_id)
    if not slots:
        return None

    idx = _round_robin_index.get(provider_id, 0)

    for offset in range(len(slots)):
        slot = slots[(idx + offset) % len(slots)]
        now = time.time()
        if slot.banned_until is not None and now < slot.banned_until:
            continue
        next_idx = (idx + offset + 1) % len(slots)
        _round_robin_index[provider_id] = next_idx
        return slot.key

    return None


def record_success(provider_id: str, key: str) -> None:
    slot = _find_slot(provider_id, key)
    if slot is None:
        return
    slot.usage_count += 1
    slot.last_used_at = time.time()
    slot.consecutive_failures = 0


def record_failure(provider_id: str, key: str) -> None:
    slot = _find_slot(provider_id, key)
    if slot is None:
        return
    slot.error_count += 1
    slot.last_used_at = time.time()
    slot.consecutive_failures += 1

    total = slot.usage_count + slot.error_count
    if total > 0 and slot.error_count / total >= MAX_ERROR_RATE:
        slot.banned_until = time.time() + BAN_DURATION_429
        logger.warning(
            f"[keys] Banned key for {provider_id} "
            f"({slot.consecutive_failures} consecutive failures, "
            f"error rate {slot.error_count / total:.0%})"
        )


def record_rate_limit(provider_id: str, key: str) -> None:
    slot = _find_slot(provider_id, key)
    if slot is None:
        return
    slot.rate_limit_hits += 1
    slot.banned_until = time.time() + BAN_DURATION_429
    logger.warning(f"[keys] Rate limited key for {provider_id}, banned 5min")


def get_key_stats(provider_id: str) -> list[dict]:
    slots = _pools.get(provider_id, [])
    return [
        {
            "usage_count": s.usage_count,
            "error_count": s.error_count,
            "last_used_at": s.last_used_at,
            "banned_until": s.banned_until,
            "rate_limit_hits": s.rate_limit_hits,
        }
        for s in slots
    ]


def all_key_pool_stats() -> dict[str, list[dict]]:
    return {pid: get_key_stats(pid) for pid in _pools}


def _find_slot(provider_id: str, key: str) -> KeySlot | None:
    slots = _pools.get(provider_id)
    if slots is None:
        return None
    for slot in slots:
        if slot.key == key:
            return slot
    return None

File: src/key-manager.ts
// ─────────────────────────────────────────────────────────────────────────────
// PrxyClaude · Key Manager  (round-robin with ban-on-429)
// ─────────────────────────────────────────────────────────────────────────────

import type { KeySlot } from "./types.js";
import { log } from "./logger.js";

const pools = new Map<string, KeySlot[]>();

export function initKeyPool(providerId: string, keys: string[]): void {
  pools.set(
    providerId,
    keys.map((key) => ({
      key,
      usageCount: 0,
      errorCount: 0,
      lastUsedAt: null,
      bannedUntil: null,
      rateLimitHits: 0,
    }))
  );
}

/** Returns the next available API key or null if all are banned. */
export function nextKey(providerId: string): string | null {
  const slots = pools.get(providerId);
  if (!slots || slots.length === 0) return null;

  const now = Date.now();
  const available = slots.filter(
    (s) => s.bannedUntil === null || s.bannedUntil <= now
  );

  if (available.length === 0) {
    // All keys are rate-limited; find the one with the soonest unban
    const soonest = slots.reduce((a, b) =>
      (a.bannedUntil ?? 0) < (b.bannedUntil ?? 0) ? a : b
    );
    const waitMs = Math.max(0, (soonest.bannedUntil ?? 0) - now);
    log("warn", `[keys] ${providerId}: all keys banned, soonest free in ${waitMs}ms`);
    return null;
  }

  // Prefer the key with the fewest recent uses
  available.sort((a, b) => a.usageCount - b.usageCount);
  const chosen = available[0];
  chosen.usageCount++;
  chosen.lastUsedAt = now;
  return chosen.key;
}

/**
 * Ban a key for a period after a 429.
 * Default: 60 s; on repeated hits, exponential up to 10 min.
 */
export function banKey(
  providerId: string,
  key: string,
  retryAfterMs = 60_000
): void {
  const slots = pools.get(providerId);
  if (!slots) return;

  const slot = slots.find((s) => s.key === key);
  if (!slot) return;

  slot.rateLimitHits++;
  const backoff = Math.min(
    retryAfterMs * Math.pow(2, slot.rateLimitHits - 1),
    600_000
  );
  slot.bannedUntil = Date.now() + backoff;
  log(
    "warn",
    `[keys] ${providerId} key ...${key.slice(-6)} banned for ${backoff / 1000}s (hit #${slot.rateLimitHits})`
  );
}

export function markKeyError(providerId: string, key: string): void {
  const slots = pools.get(providerId);
  const slot = slots?.find((s) => s.key === key);
  if (slot) slot.errorCount++;
}

export function getKeyStats(providerId: string): Omit<KeySlot, "key">[] {
  const slots = pools.get(providerId);
  if (!slots) return [];
  return slots.map(({ key, ...rest }) => rest); // don't leak keys
}

export function allKeyPoolStats(): Record<string, ReturnType<typeof getKeyStats>> {
  const out: Record<string, ReturnType<typeof getKeyStats>> = {};
  for (const [id] of pools) out[id] = getKeyStats(id);
  return out;
}
